network/gossip/libp2p/network.go

```diff
@@ -156,38 +172,48 @@ func (n *Network) Identity() (map[flow.Identifier]flow.Identity, error) {
 }
 
 // Topology returns the identities of a uniform subset of nodes in protocol state using the topology provided earlier
-func (n *Network) Topology() (map[flow.Identifier]flow.Identity, error) {
-	subset, err := n.top.Subset(n.ids, n.fanout())
+func (n *Network) Topology() (flow.IdentityList, error) {
+	fanout := n.fanout()
+	n.RLock()
+	defer n.RUnlock()
+	subset, err := n.top.Subset(n.ids, fanout)
 	if err != nil {
 		return nil, fmt.Errorf("failed to derive list of peer nodes to connect to: %w", err)
 	}
-
-	// creates a map of all the selected ids
-	topMap := make(map[flow.Identifier]flow.Identity)
-	for _, id := range subset {
-		topMap[id.NodeID] = *id
-	}
-	return topMap, nil
+	return subset, nil
 }
 
 func (n *Network) Receive(nodeID flow.Identifier, msg *message.Message) error {
-
 	err := n.processNetworkMessage(nodeID, msg)
 	if err != nil {
 		return fmt.Errorf("could not process message: %w", err)
 	}
-
 	return nil
 }
 
-func (n *Network) SetIDs(ids flow.IdentityList) {
-	// remove this node id from the list of fanout target ids to avoid self-dial
-	idsMinusMe := ids.Filter(n.me.NotMeFilter())
-	n.ids = idsMinusMe
+// SetIDs updates the identity list cached by the network layer
+func (n *Network) SetIDs(ids flow.IdentityList) error {
+
+	// remove self from id
+	ids = ids.Filter(n.me.NotMeFilter())
+
+	n.Lock()
+	n.ids = ids
+	n.Unlock()
+
+	// update the allow list
+	err := n.mw.UpdateAllowList()
```

Comment:
shouldn't we also cover this method with a `Lock`?

Reply:
`UpdateAllowList` calls back `network.Identity` which in turn does a `lock`. Hence, it would end up in deadlock if I covered `UpdateAllowList` with it.